internal/testutils/broker.go

```diff
@@ -0,0 +1,326 @@
+func (b *BrokerBusMock) NewSession(username, lang string) (sessionID, encryptionKey string, dbusErr *dbus.Error) {
+	if username == "NS_bus_error" {
+		return "", "", dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	sessionID, encryptionKey, err := b.broker.NewSession(context.Background(), username, lang)
+	if err != nil {
+		return "", "", dbus.MakeFailedError(err)
+	}
+	return sessionID, encryptionKey, nil
+}
+
+// GetAuthenticationModes calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) GetAuthenticationModes(sessionID string, supportedUILayouts []map[string]string) (authenticationModes []map[string]string, dbusErr *dbus.Error) {
+	if sessionID == "GAM_bus_error" {
+		return nil, dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	authenticationModes, err := b.broker.GetAuthenticationModes(context.Background(), sessionID, supportedUILayouts)
+	if err != nil {
+		return nil, dbus.MakeFailedError(err)
+	}
+	return authenticationModes, nil
+}
+
+// SelectAuthenticationMode calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) SelectAuthenticationMode(sessionID, authenticationModeName string) (uiLayoutInfo map[string]string, dbusErr *dbus.Error) {
+	if sessionID == "SAM_bus_error" {
+		return nil, dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	uiLayoutInfo, err := b.broker.SelectAuthenticationMode(context.Background(), sessionID, authenticationModeName)
+	if err != nil {
+		return nil, dbus.MakeFailedError(err)
+	}
+	return uiLayoutInfo, nil
+}
+
+// IsAuthorized calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) IsAuthorized(sessionID, authenticationData string) (access, infoUser string, dbusErr *dbus.Error) {
+	if sessionID == "IA_bus_error" {
+		return "", "", dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	access, infoUser, err := b.broker.IsAuthorized(context.Background(), sessionID, authenticationData)
+	if err != nil {
+		return "", "", dbus.MakeFailedError(err)
+	}
+	return access, infoUser, nil
+}
+
+// EndSession calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) EndSession(sessionID string) (dbusErr *dbus.Error) {
+	if sessionID == "ES_bus_error" {
+		return dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	err := b.broker.EndSession(context.Background(), sessionID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// CancelIsAuthorized calls the broker mock corresponding method.
+func (b *BrokerBusMock) CancelIsAuthorized(sessionID string) (dbusErr *dbus.Error) {
+	b.broker.CancelIsAuthorized(context.Background(), sessionID)
+	return nil
+}
+
+/* BrokerMock implementation */
```

Comment:
This has two reasons:
1) In order to implement the [brokerer interface], our functions must have a `context.Context` argument as well (and this is not possible in an exported dbus function, since contexts can not be reflected), so the `BrokerBusMock` is basically a translator of calls to the actual broker (`BrokerMock`);
2) Implementing it this way, we can use the `BrokerMock` both "onbus" and "offbus" since the call handling is made by the `BrokerBusMock` (very similar to what I did in the [examplebroker]()). This was not needed in the unit tests, but it could prove itself very useful in the integration tests.

Reply:
This is right, but I feel this is overcomplifies for little benefit:
- there is no need to explicit have a "broker" object on brokerbusmock which implements the brokerer interface, as this is only on the caller side that this interface is used
- I don’t see any value of having offbus calls. What we want to validate here is the whole chain, using the dbusbroker middleware. So, I think merging the 2 types is what is making the most sense for long term support.